Naming a constructor `PS` breaks the javascript

mhcurylo opened the pull request:
I have fixed the PS constructor braking the exports by changing the IIFE argument from exports to the PS module object.
There are no automated tests for the psc-bundle and I have not added one. Maybe they should be introduced?
Fragment of an output is:
```
function(PS) {
  // Generated by purs version 0.12.2
  "use strict";
  var exports = PS["Effect.Console"] = PS["Effect.Console"] || {};
  var $foreign = PS["Effect.Console"];
  var Data_Show = PS["Data.Show"];
  var Data_Unit = PS["Data.Unit"];
  var Effect = PS["Effect"];
  exports["log"] = $foreign.log;
})(PS); 
```

hdgarrood commented:
I agree that having some tests would be good. Are you volunteering to add some? I'm very happy to help if you have any questions.

mhcurylo commented:
It sounds like morning fun if we do not mind me taking some time. An outline:
1. Example directory for purs files in tests/purs/bundle directory.
2. A module TestPscBundle.hs in /tests/ directory.
3. Similar setup as the compiler tests but using the psc-bundle.
As an alternative, we could add an annotation 'use-bundle' to the compiler tests and keep the examples in there. I think it might be mixing concerns though.

hdgarrood commented:
I think a tests/purs/bundle directory sounds great.

hdgarrood commented:
(Fixes #3505)

mhcurylo commented:
@hdgarrood 
Test added. 
I have exported the common functions between compiler and bundle tests to TestUtils.
One problem I see is that in order to increase bundle coverage we would have to manually implement js scenarios. With the current solution (test vs compiled purescript code) - some code will be unreachable.
For now, it fixes the problem and adds coverage. 
Please review.

hdgarrood commented on tests/TestPscBundle.hs:
```diff
@@ -0,0 +1,97 @@
+import Test.Tasty.Hspec
+
+main :: IO TestTree
+main = testSpec "psc-bundle" spec
```
Can you use just "bundle" here? `psc-bundle` doesn't exist any more.

mhcurylo replied:
Should I rename the module and the spec as well, then?

hdgarrood replied:
Yes please!

hdgarrood commented on tests/TestPscBundle.hs:
```diff
@@ -0,0 +1,97 @@
+      Left errs -> return . Just . P.prettyPrintMultipleErrors P.defaultPPEOptions $ errs
+      Right _ -> do
+        process <- findNodeProcess
+        -- Mateuszu! Tutaj zrob. Trzeba to zbundlowac zamiast tworzyc Main.main()
```
If this comment should stay in, can you write it in English please?

mhcurylo replied:
This comment was written in Polish, to mark WIP, so that I never ever forget to delete that one and the ones next to it.

hdgarrood commented on tests/TestPscBundle.hs:
```diff
@@ -0,0 +1,97 @@
+      Right _ -> do
+        process <- findNodeProcess
+        -- Mateuszu! Tutaj zrob. Trzeba to zbundlowac zamiast tworzyc Main.main()
+        --app modulesDir entryPoint
```
Can this comment be removed?

hdgarrood commented on tests/TestPscBundle.hs:
```diff
@@ -0,0 +1,97 @@
+        case bundled of
+            Right (_, js) -> do
+              writeUTF8File entryPoint js
+              --writeFile entryPoint "require('Main').main()"
```
Can this comment be removed?

hdgarrood commented on tests/TestPscBundle.hs:
```diff
@@ -0,0 +1,97 @@
+            Left err -> return . Just $ "Coud not bundle: " ++ show err
+
+logfile :: FilePath
+logfile = "psc-bundle-tests.out"
```
Same here re "psc-bundle"

hdgarrood commented on tests/support/pscide/purs.tix:
```diff
@@ -0,0 +1 @@
+Tix [ TixModule "purescript-0.12.2-KMJAZ5OYzDhGVgSCJVsyOl/Language.PureScript.TypeChecker.Subsumption" 1527027485 174 [1,0,12,1,1,1,1,1,1,1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0 ...
```
Can you remove this file please?

hdgarrood commented on tests/Main.hs:
```diff
@@ -43,8 +45,10 @@ main = do
     testGroup
       "Tests"
-      [ compilerTests
+      [ 
```
Would you remind removing the extra line here?

hdgarrood commented:
> One problem I see is that in order to increase bundle coverage we would have to manually implement js scenarios. With the current solution (test vs compiled purescript code) - some code will be unreachable.

This doesn't bother me, I think what we have here is fine for now.

mhcurylo commented:
Thanks for all the minor comments, they are applied.

hdgarrood commented:
I've now reviewed this properly and it looks good to me, although I have one small suggestion. Instead of e.g.
```
  var exports = PS["Effect.Console"] = PS["Effect.Console"] || {};
```
could we perhaps generate
```
  PS["Effect.Console"] = PS["Effect.Console"] || {};
  var exports = PS["Effect.Console"];
```
which I think should be equivalent, but is a bit clearer?

hdgarrood commented:
Oh and one more: how about using $PS as the name of the argument for the IIFE function for each module in the output? Then we should be able to avoid collisions with names from PureScript values entirely, rather than relying on slightly arcane details of JS scoping in cases of collisions.

mhcurylo commented:
All suggestions included in the PR.
I have changed the generated JS. 
I have changed the default namespace to $PS (inside the bundle action in the app) as suggested.
It would not solve the issue by itself - a namespace provided to the bundler could still interfere with a constructor if they shared names.
Thus I have changed the name of the test and included a pragma to configure namespace provided to the bundler so that the test is still testing what it should (the case where the namespace is the same as the constructor).

hdgarrood commented:
Sorry, I should have been clearer: what I mean is that the default bundle namespace should continue to be PS, but the name given to the argument of each iife for each module should be something which couldn’t clash with any purescript name. So for example you might get

```javascript
var PS = PS || {};
function($PS) {
  // Generated by purs version 0.12.2
  "use strict";
  var exports = $PS["Effect.Console"] = $PS["Effect.Console"] || {};
  var $foreign = $PS["Effect.Console"];
  var Data_Show = $PS["Data.Show"];
  var Data_Unit = $PS["Data.Unit"];
  var Effect = $PS["Effect"];
  exports["log"] = $foreign.log;
})(PS);
```

So the namespace object is always referred to as $PS inside each iife, regardless of the namespace CLI option. I’d prefer to leave the tests as they were before.

hdgarrood commented on src/Language/PureScript/Bundle.hs:
```diff
@@ -638,17 +644,20 @@ codeGen optionsMainModule optionsNamespace ms outFileOpt = (fmap sourceMapping o
+      addModuleExports :: [JSStatement] -> [JSStatement]
+      addModuleExports [] = lfHead moduleExports 
+      addModuleExports (x:xs) = lfHead [x] ++ moduleExports ++ xs
+      moduleExports = [JSExpressionStatement (JSAssignExpression (innerModuleReference lfsp mn) (JSAssign sp) (JSExpressionBinary (innerModuleReference sp mn) (JSBinOpOr sp) (emptyObj sp))) (JSSemi JSAnnotSpace), JSVariable lfsp (JSLOne $ JSVarInitExpression (JSIdentifier sp "exports") $ JSVarInit sp (innerModuleReference sp mn)) $ JSSemi JSNoAnnot ]
```
Last nitpick and then I think this is good to merge: this line is too long. Can you please put each element of this list on its own line and use indentation similar to how it has been used elsewhere in this file (such as in the main body of `wrap` just above here)?

mhcurylo commented:
Oh, that sounds reasonable.
All changes are reverted/made.

The only difference to the example is in the namespace declaration.
Instead of the proposed
```
var PS = PS || {};
```
it is
```
var PS = {};
```
as it was before.
The semantics are different when a namespace is shared across bundles.
Which one should we go with?

hdgarrood commented:
Yes, thanks, I think the initial namespace declaration should stay as it is, `var PS = {};`. (I couldn't be bothered to check how it is currently and so I made a guess which obviously has turned out to be incorrect.)

Did you make a commit locally which you haven't pushed yet? It looks like there aren't any new commits to this branch since my last review comment.

mhcurylo commented:
I have rebased and pushed.
Everything is up to date.

hdgarrood commented:
The `moduleExports` declaration doesn't seem to have changed?

mhcurylo commented:
Thanks for noticing.
Somewhat I did not read the moduleExports comment.
Now it is addressed.

mhcurylo commented:
Thanks!
